In [1]:
from pandas import Series, DataFrame
import pandas as pd
%pylab inline

# reading in file and creating headers for the dataframe
births_raw = pd.read_csv("C:/Users/Jason Nania/Desktop/UT Classes/Data Files/US_births(2018).csv")

%pylab is deprecated, use %matplotlib inline and import the required libraries.
Populating the interactive namespace from numpy and matplotlib


C:\Users\Jason Nania\AppData\Local\Temp\ipykernel_11848\1742040395.py:6: DtypeWarning: Columns (7) have mixed types. Specify dtype option on import or set low_memory=False.
  births_raw = pd.read_csv("C:/Users/Jason Nania/Desktop/UT Classes/Data Files/US_births(2018).csv")


In [2]:
# Giving the columns usable names instead of abbreviations
births_raw.columns = ['Attendant_at_Birth', 'Birth_Place','Body_Mass_Index','Cigarettes_Before_Pregnancy','Birth_Weigth_Grams',
         'Last_Normal_Menses_Month','Last_Normal_Menses Year','Marital_Status','Birth_Month','Time_of_Birth',
         'Birth_Day_of_Week','Birth_Year','Delivery_Weight','Father_Age','Father_Education','Father_Hispanic_Origin',
         'Father_Race_Recode_15','Father_Race_Recode_31','Father_Race_Recode_6','Interval_Since_Last_Live_Birth_Recode',
         'Interval_Since_Last_Other_Pregnancy','Interval_Since_Last_Pregnancy_Recode','Imputed_Sex','Gonorrhea',
         'Induction_of_Labor','Mother_Age','Mother_Age_Imputed','Mother_Marital_Status_Imputed','Mother_Nativity',
         'Mother_Education','Mother_Hispanic_Origin','Admit_to_Intensive_Care','Mother_Race_Recode_15','Mother_Race_Recode_31',
         'Mother_Race_Imputed_Flag','Mother_Race_Recode_6','Mother_Transferred','Mother_Height_Inches','No_Infections_Reported',
         'No_Maternal_Morbidity_Reported','No_Risk_Factors_Reported','Payment_Source_for_Delivery','Payment_Recode',
         'Month_Prenatal_Care_Began','Number_of_Prenatal_Visits','Prior_Births_Now_Dead','Prior_Births_Now_Living',
         'Prior_Other_Terminations','Pre_pregnancy_Weight_lbs','Delivery_Method_Recode','Residence_Status','Previous_Cesarean',
         'Number_of_Previous_Cesareans','Sex_of_Infant','Weight_Gain']
births = births_raw.drop(['Mother_Age_Imputed','Mother_Marital_Status_Imputed','Mother_Race_Imputed_Flag','Imputed_Sex'], axis = 1)
births

Attendant_at_Birth  Birth_Place  Body_Mass_Index  \
0                         1            1             30.7   
1                         1            1             33.3   
2                         1            1             30.0   
3                         3            1             23.7   
4                         1            1             35.5   
...                     ...          ...              ...   
3801529                   1            1             26.3   
3801530                   1            1             32.4   
3801531                   5            7             27.6   
3801532                   5            4             99.9   
3801533                   1            1             42.1   

         Cigarettes_Before_Pregnancy  Birth_Weigth_Grams  \
0                                  0                3657   
1                                  2                3242   
2                                  0                3470   
3                                  0                3140   
4                                  0                2125   
...                              ...                 ...   
3801529                            0                2140   
3801530                            0                3560   
3801531                            0                3367   
3801532                           98                 320   
3801533                            0                2455   

         Last_Normal_Menses_Month  Last_Normal_Menses Year Marital_Status  \
0                               4                     2017              1   
1                              99                     9999              2   
2                               4                     2017              1   
3                               5                     2017              2   
4                              99                     9999              1   
...                           ...                      ...            ...   
3801529                         4                     2018              2   
3801530                         3                     2018              2   
3801531                         3                     2018              2   
3801532                         3                     2018              2   
3801533                         2                     2018              1   

         Birth_Month  Time_of_Birth  ...  Prior_Births_Now_Dead  \
0                  1           1227  ...                      0   
1                  1           1704  ...                      0   
2                  1            336  ...                      0   
3                  1            938  ...                      0   
4                  1            830  ...                      0   
...              ...            ...  ...                    ...   
3801529           12           1531  ...                      0   
3801530           12           2347  ...                      0   
3801531           11           1354  ...                      0   
3801532            8            859  ...                      0   
3801533           11           1018  ...                      0   

         Prior_Births_Now_Living  Prior_Other_Terminations  \
0                              1                         2   
1                              2                         0   
2                              1                         0   
3                              2                         0   
4                              1                         4   
...                          ...                       ...   
3801529                        1                         0   
3801530                        2                         1   
3801531                        5                         1   
3801532                        0                         0   
3801533                        0                         0   

         Pre_pregnancy_Weight_lbs  Delivery_Method_Recode  Residence_Status  \
0

In [3]:
# replacing strings with integers in the Marital Status column
births['Marital_Status'] = births['Marital_Status'].replace(regex='1', value=1)
births['Marital_Status'] = births['Marital_Status'].replace(regex='2', value=2)

# replacing blanks with NaN
births = births.replace(regex=' ', value=np.nan)
births.dtypes

Attendant_at_Birth                         int64
Birth_Place                                int64
Body_Mass_Index                          float64
Cigarettes_Before_Pregnancy                int64
Birth_Weigth_Grams                         int64
Last_Normal_Menses_Month                   int64
Last_Normal_Menses Year                    int64
Marital_Status                           float64
Birth_Month                                int64
Time_of_Birth                              int64
Birth_Day_of_Week                          int64
Birth_Year                                 int64
Delivery_Weight                            int64
Father_Age                                 int64
Father_Education                           int64
Father_Hispanic_Origin                     int64
Father_Race_Recode_15                      int64
Father_Race_Recode_31                      int64
Father_Race_Recode_6                       int64
Interval_Since_Last_Live_Birth_Recode      int64
Interval_Since_Last_

In [4]:
# checking values of Marital Status are not strings
births['Marital_Status'].value_counts()

1.0    2008022
2.0    1337429
Name: Marital_Status, dtype: int64

In [5]:
# checking for columns with multiple data types
# for col in births.columns:
#     weird = (births[[col]].applymap(type) != births[[col]].iloc[0].apply(type)).any(axis=1)
#     if len(births[weird]) > 0:
#         print(col)

In [6]:
births['No_Infections_Reported'].value_counts()

1    3688228
0     104732
9       8574
Name: No_Infections_Reported, dtype: int64

In [7]:
# Logistic Regression for Infections
from patsy import dmatrices
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
import warnings
warnings.filterwarnings('ignore')
# replacing 9 value with NaN. 9 indicates unknown
births['No_Infections_Reported'] = births['No_Infections_Reported'].replace(9, value=np.nan)
births['No_Infections_Reported'].value_counts()


formula = 'No_Infections_Reported ~ 0 + C(Attendant_at_Birth) + C(Birth_Place) + Body_Mass_Index + Cigarettes_Before_Pregnancy' + \
           ' + Birth_Weigth_Grams + C(Marital_Status) + Delivery_Weight + C(Induction_of_Labor) + Mother_Age + Admit_to_Intensive_Care' +\
            '+ Month_Prenatal_Care_Began + Number_of_Prenatal_Visits + Prior_Births_Now_Living + Pre_pregnancy_Weight_lbs' +\
            '+ C(Delivery_Method_Recode) + Previous_Cesarean + Number_of_Previous_Cesareans + Sex_of_Infant'

# setting up the matrix
Y, X = dmatrices(formula, births, return_type='dataframe')
y = Y['No_Infections_Reported'].values

# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.4, random_state=1)

# running the logistic regression
model = LogisticRegression()
result = model.fit(X_train, y_train)

# using model to predict train
prediction_train = model.predict(X_train)
print(metrics.accuracy_score(y_train, prediction_train))

# using model to predict test
prediction_test = model.predict(X_test)
print(metrics.accuracy_score(y_test, prediction_test))

# feature coefficients
weights = Series(model.coef_[0], # model.coef_[0] is necessary because the output is a  list, and this is how to access it
                 index=X.columns.values)
weights.sort_values()



<frozen importlib._bootstrap>:228: RuntimeWarning: scipy._lib.messagestream.MessageStream size changed, may indicate binary incompatibility. Expected 56 from C header, got 64 from PyObject
C:\Users\Jason Nania\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


0.970368942280222
0.9701469614286186


Cigarettes_Before_Pregnancy      -1.877566e-02
C(Marital_Status)[T.2.0]         -1.377764e-02
Month_Prenatal_Care_Began        -6.026729e-03
Prior_Births_Now_Living          -3.701560e-03
Number_of_Previous_Cesareans     -1.576730e-03
C(Induction_of_Labor)[T.Y]       -1.255596e-03
Pre_pregnancy_Weight_lbs         -3.894238e-04
C(Attendant_at_Birth)[2]         -1.950025e-04
C(Attendant_at_Birth)[3]         -1.691507e-04
Delivery_Weight                  -6.817717e-05
C(Induction_of_Labor)[T.U]       -3.851871e-05
C(Birth_Place)[T.4]              -3.357942e-05
C(Delivery_Method_Recode)[T.5]   -1.573813e-05
Previous_Cesarean[T.U]           -1.280180e-05
Admit_to_Intensive_Care[T.Y]     -8.105406e-06
Admit_to_Intensive_Care[T.U]     -7.132991e-06
C(Delivery_Method_Recode)[T.9]   -3.059701e-06
C(Birth_Place)[T.9]              -3.846578e-07
C(Birth_Place)[T.7]               1.164839e-06
C(Attendant_at_Birth)[9]          1.222608e-06
C(Birth_Place)[T.5]               1.361019e-06
C(Delivery_Me

In [10]:
from sklearn import tree
# creating tree model
modelT = tree.DecisionTreeClassifier(criterion='entropy')

resultT = modelT.fit(X_train, y_train)

prediction_trainT = modelT.predict(X_train)
print(metrics.accuracy_score(y_train, prediction_trainT))

prediction_testT = model.predict(X_test)
print(metrics.accuracy_score(y_test, prediction_testT))


# creating max depth tree
modelT2 = tree.DecisionTreeClassifier(criterion='entropy', max_depth=2)
resultT2 = modelT2.fit(X_train, y_train)

prediction_trainT2 = modelT2.predict(X_train)
print(metrics.accuracy_score(y_train, prediction_trainT2))

prediction_testT2 = modelT2.predict(X_test)
print(metrics.accuracy_score(y_test, prediction_testT2))

1.0
0.9701469614286186
0.9703699412033827
0.9701499581951075


In [21]:
w_sort = weights.sort_values()
w_sort

Cigarettes_Before_Pregnancy      -1.877566e-02
C(Marital_Status)[T.2.0]         -1.377764e-02
Month_Prenatal_Care_Began        -6.026729e-03
Prior_Births_Now_Living          -3.701560e-03
Number_of_Previous_Cesareans     -1.576730e-03
C(Induction_of_Labor)[T.Y]       -1.255596e-03
Pre_pregnancy_Weight_lbs         -3.894238e-04
C(Attendant_at_Birth)[2]         -1.950025e-04
C(Attendant_at_Birth)[3]         -1.691507e-04
Delivery_Weight                  -6.817717e-05
C(Induction_of_Labor)[T.U]       -3.851871e-05
C(Birth_Place)[T.4]              -3.357942e-05
C(Delivery_Method_Recode)[T.5]   -1.573813e-05
Previous_Cesarean[T.U]           -1.280180e-05
Admit_to_Intensive_Care[T.Y]     -8.105406e-06
Admit_to_Intensive_Care[T.U]     -7.132991e-06
C(Delivery_Method_Recode)[T.9]   -3.059701e-06
C(Birth_Place)[T.9]              -3.846578e-07
C(Birth_Place)[T.7]               1.164839e-06
C(Attendant_at_Birth)[9]          1.222608e-06
C(Birth_Place)[T.5]               1.361019e-06
C(Delivery_Me

# Accuracy
* Baseline (predict all 1) = 0.9724
* Logistic Regression = 0.97016
* Decision Tree (no max depth) = 0.97015
* Decision Tree (max depth = 2) = 0.97015

# Logistic Regression Coefficients
* Cigarettes_Before_Pregnancy      -1.877566e-02
* C(Marital_Status)[T.2.0]         -1.377764e-02
* Month_Prenatal_Care_Began        -6.026729e-03
* Number_of_Previous_Cesareans     -1.576730e-03
* C(Induction_of_Labor)[T.Y]       -1.255596e-03


https://journals.plos.org/plosmedicine/article?id=10.1371/journal.pmed.1003929 - marriage does lead to lower infection rates

# Conclusion
We are not able to beat the accuracy of the baseline where all births are assumed to have no infection. However, we can use our model to identify those patients who may be at a higher-risk of pregnancy using the logistic regression model. This can lead to proactive action during the delivery and immediately thereafter to try and prevent infection. This can be especially practical if medical professionals can identify any of the variables that are the largest contributors to increasing the chance of an infeciton. These variables are Cigarettes before pregnancy, marital status, month prenatal care began, number of previous cesareans, and induction of labor. The most interesting variable in this group is the marital status because it is not necessarily intuitive that this should matter when predicting whether an infection will occur during birth. As noted in the article by St.Germain et al. there is confirmed evidence that if a mother is unmarried there is a higher chance of infection during birth. 